src/gui/dialogs/units_dialog.cpp

```diff
@@ -0,0 +1,507 @@
+	}
+}
+
+std::string get_title_suffix(int side_num)
+{
+	if(!resources::gameboard) {
+		return "";
+	}
+
+	unit_map& units = resources::gameboard->units();
+
+	int controlled_recruiters = 0;
+	for(const auto& team : resources::gameboard->teams()) {
+		if(team.is_local_human() && !team.recruits().empty() && units.find_leader(team.side()) !=units.end()) {
+			++controlled_recruiters;
+		}
+	}
+
+	std::stringstream msg;
+	if(controlled_recruiters >= 2) {
+		unit_map::const_iterator leader = resources::gameboard->units().find_leader(side_num);
+		if(leader != resources::gameboard->units().end() && !leader->name().empty()) {
+			msg << " (" << leader->name() << ")";
+		}
+	}
+
+	return msg.str();
+}
+}
+
+void units_dialog::pre_show()
+{
+	label& title = find_widget<label>("title", true);
+	if (team_) {
+		title.set_label(title.get_label() + get_title_suffix(team_->side()));
+	}
+
+	text_box& filter = find_widget<text_box>("filter_box");
+	connect_signal_notify_modified(filter, std::bind(&units_dialog::filter_text_changed, this));
+
+	listbox& list = find_widget<listbox>("recall_list");
+	connect_signal_notify_modified(list, std::bind(&units_dialog::list_item_clicked, this));
+
+	connect_signal_mouse_left_click(
+		find_widget<button>("show_help"),
+		std::bind(&units_dialog::show_help, this));
+
+	list.clear();
+
+	keyboard_capture(&filter);
+	add_to_keyboard_chain(&list);
+
+	if (show_gender_grid_) {
+		toggle_button& male_toggle = find_widget<toggle_button>("male_toggle");
+		toggle_button& female_toggle = find_widget<toggle_button>("female_toggle");
+
+		gender_toggle_.add_member(&male_toggle, unit_race::MALE);
+		gender_toggle_.add_member(&female_toggle, unit_race::FEMALE);
+		gender_toggle_.set_member_states(unit_race::MALE);
+		gender_toggle_.set_callback_on_value_change(
+			std::bind(&units_dialog::gender_toggle_callback, this, std::placeholders::_2));
+	}
+
+	if (show_variation_grid_) {
+		menu_button& var_box = find_widget<menu_button>("variation_box");
+		connect_signal_notify_modified(var_box, std::bind(&units_dialog::variation_menu_callback, this));
+	}
+
+	if (show_rename_) {
+		connect_signal_mouse_left_click(
+		find_widget<button>("rename"),
+		std::bind(&units_dialog::rename_unit, this));
+	}
+
+	if (show_dismiss_ && team_) {
+		connect_signal_mouse_left_click(
+			find_widget<button>("dismiss"),
+			std::bind(&units_dialog::dismiss_unit, this));
+	}
+
+	show_list(list);
+
+	find_widget<label>("title").set_label(title_);
+	find_widget<button>("ok").set_label(ok_label_);
+	find_widget<button>("dismiss").set_visible(
+		show_dismiss_ ? widget::visibility::visible : widget::visibility::invisible);
+	find_widget<button>("rename").set_visible(
+		show_rename_ ? widget::visibility::visible : widget::visibility::invisible);
+	find_widget<grid>("variation_gender_grid").set_visible(
+		show_gender_grid_ ? widget::visibility::visible : widget::visibility::invisible);
+
+	for (int i = 0; i <= 7; i++) {
```

Comment:
Aren't there 8 columns?

Reply:
the condition is `<=7`, but i'd get it from `visible_headers_`

Reply:
updated to use `visible_headers_.size()`